crates/containerd-shim-wasmedge/src/executor.rs

```diff
@@ -22,45 +24,45 @@ impl WasmEdgeExecutor {
 
 impl Executor for WasmEdgeExecutor {
     fn exec(&self, spec: &Spec) -> Result<(), ExecutorError> {
-        // parse wasi parameters
-        let args = oci::get_args(spec);
-        if args.is_empty() {
-            return Err(ExecutorError::InvalidArg);
-        }
+        match self.can_handle(spec) {
+            Ok(()) => {
+                // parse wasi parameters
+                let args = oci::get_args(spec);
+                if args.is_empty() {
+                    return Err(ExecutorError::InvalidArg);
+                }
 
-        let vm = self
-            .prepare(args, spec)
-            .map_err(|err| ExecutorError::Other(format!("failed to prepare function: {}", err)))?;
-
-        // TODO: How to get exit code?
-        // This was relatively straight forward in go, but wasi and wasmtime are totally separate things in rust
-        let (module_name, method) = oci::get_module(spec);
-        debug!("running {:?} with method {}", module_name, method);
-        match vm.run_func(Some("main"), method, params!()) {
-            Ok(_) => std::process::exit(0),
-            Err(_) => std::process::exit(137),
-        };
-    }
+                let vm = self.prepare(args, spec).map_err(|err| {
+                    ExecutorError::Other(format!("failed to prepare function: {}", err))
+                })?;
 
-    fn can_handle(&self, spec: &Spec) -> bool {
-        // check if the entrypoint of the spec is a wasm binary.
-        let (module_name, _method) = oci::get_module(spec);
-        let module_name = match module_name {
-            Some(m) => m,
-            None => {
-                log::info!("WasmEdge cannot handle this workload, no arguments provided");
-                return false;
+                // TODO: How to get exit code?
+                // This was relatively straight forward in go, but wasi and wasmtime are totally separate things in rust
+                let (module_name, method) = oci::get_module(spec);
+                debug!("running {:?} with method {}", module_name, method);
+                match vm.run_func(Some("main"), method, params!()) {
+                    Ok(_) => std::process::exit(0),
+                    Err(_) => std::process::exit(137),
+                };
             }
-        };
-        let path = PathBuf::from(module_name);
-
-        path.extension()
-            .map(|ext| ext.to_ascii_lowercase())
-            .is_some_and(|ext| ext == "wasm" || ext == "wat")
+            Err(ExecutorValidationError::CantHandle(_)) => {
+                LinuxContainerExecutor::new(self.stdio.clone()).exec(spec)?;
+                Ok(())
+            }
+            Err(_) => Err(ExecutorError::InvalidArg),
+        }
     }
 
-    fn name(&self) -> &'static str {
-        EXECUTOR_NAME
+    fn validate(&self, spec: &Spec) -> std::result::Result<(), ExecutorValidationError> {
+        match self.can_handle(spec) {
```

Comment:
**nit:** I'm not a huge fan of doing this check twice, here and on `exec`. But it works.
I'll try to think if we can avoid it or not.

Reply:
This is one of the downside/trade off of this interface. In both valid and exec, we have to run the `can_handle` logic to figure out the correct exec. Also due to rust safety features, in `exec` we have to run a lot of the similar checks to return `InvalidArg` error. For example, if there is no process block in the spec, we have to check it again inside `exec` even though we already checked in the `valid`. It doesn't make sense for us to `panic`.